switch/ffmpeg/tls.patch

```diff
@@ -0,0 +1,273 @@
++ * Lesser General Public License for more details.
++ *
++ * You should have received a copy of the GNU Lesser General Public
++ * License along with FFmpeg; if not, write to the Free Software
++ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
++ */
++
++#include <switch.h>
++
++#include "avformat.h"
++#include "internal.h"
++#include "url.h"
++#include "tls.h"
++#include "libavutil/parseutils.h"
++
++typedef struct TLSContext {
++    const AVClass *class;
++    TLSShared tls_shared;
++    SslContext context;
++    SslConnection conn;
++} TLSContext;
++
++static int tls_close(URLContext *h)
++{
++    TLSContext *tls_ctx = h->priv_data;
++
++    sslConnectionClose(&tls_ctx->conn);
++    sslContextClose(&tls_ctx->context);
++
++    ffurl_closep(&tls_ctx->tls_shared.tcp);
++    return 0;
++}
++
++static int tls_open(URLContext *h, const char *uri, int flags, AVDictionary **options)
++{
++    TLSContext *tls_ctx = h->priv_data;
++    TLSShared *shr = &tls_ctx->tls_shared;
++    int ret;
++    Result rc;
++
++    if ((ret = ff_tls_open_underlying(shr, h, uri, options)) < 0)
++        goto fail;
++
++    if (shr->listen) {
++        av_log(h, AV_LOG_ERROR, "TLS Listen Sockets with libnx is not implemented.\n");
++        ret = AVERROR(EINVAL);
++        goto fail;
++    }
++
++    rc = sslCreateContext(&tls_ctx->context, SslVersion_Auto);
++    if (R_FAILED(rc)) {
++        ret = AVERROR(EINVAL);
++        goto fail;
++    }
++
++    rc = sslContextCreateConnection(&tls_ctx->context, &tls_ctx->conn);
++
++    if (R_SUCCEEDED(rc)) {
++        rc = sslConnectionSetOption(&tls_ctx->conn, SslOptionType_DoNotCloseSocket, true);
++    }
++
++    if (R_SUCCEEDED(rc)) {
++        int *sockfd = shr->tcp->priv_data + sizeof(const AVClass *);
++        ret = socketSslConnectionSetSocketDescriptor(&tls_ctx->conn, *sockfd);
++        if (ret == -1 && errno != ENOENT) {
++            av_log(h, AV_LOG_ERROR, "socketSslConnectionSetSocketDescriptor returned: %u\n", rc);
++            goto fail;
++        }
++    }
++
++    if (R_SUCCEEDED(rc)) {
++        rc = sslConnectionSetHostName(&tls_ctx->conn, shr->host, strlen(shr->host));
++    }
++
++    /* This will fail on system-versions where this option isn't available,
++     * so ignore errors from this. */
++    if (R_SUCCEEDED(rc)) {
++        sslConnectionSetOption(&tls_ctx->conn, SslOptionType_SkipDefaultVerify, true);
++    }
++
++    if (R_SUCCEEDED(rc) && hosversionAtLeast(3, 0, 0)) {
++        rc = sslConnectionSetOption(&tls_ctx->conn, SslOptionType_GetServerCertChain, true);
++    }
++
++    if (R_SUCCEEDED(rc)) {
++        u32 verifyopt = SslVerifyOption_DateCheck;
++        if (shr->verify) verifyopt |= SslVerifyOption_PeerCa;
++        rc = sslConnectionSetVerifyOption(&tls_ctx->conn, verifyopt);
++        if (R_FAILED(rc)) goto fail;
++    }
++
++    while (rc = sslConnectionDoHandshake(&tls_ctx->conn, NULL, NULL, NULL, 0)) {
```

Comment:
While-loop is not needed when using the default IoMode which is blocking. Since you're not using server-cert, SslOptionType_GetServerCertChain above can be removed.

Reply:
@yellows8 removed `SslOptionType_GetServerCertChain` and add `sslConnectionSetIoMode`